homeassistant/helpers/config_validation.py

```diff
@@ -518,18 +535,90 @@ def ensure_list_csv(value: Any) -> Sequence:
     return ensure_list(value)
 
 
-def deprecated(key):
-    """Log key as deprecated."""
+def deprecated(key: str,
+               replacement_key: Optional[str] = None,
+               invalidation_version: Optional[str] = None,
+               default: Optional[Any] = None):
+    """
+    Log key as deprecated and provide a replacement (if exists).
+
+    Expected behavior:
+        - Outputs the appropriate deprecation warning if key is detected
+        - Processes schema moving the value from key to replacement_key
+        - Processes schema changing nothing if only replacement_key provided
+        - No warning if only replacement_key provided
+        - No warning if neither key nor replacement_key are provided
+            - Adds replacement_key with default value in this case
+        - Once the invalidation_version is crossed, raises vol.Invalid if key
+        is detected
+    """
     module_name = inspect.getmodule(inspect.stack()[1][0]).__name__
 
-    def validator(config):
+    if replacement_key and invalidation_version:
+        warning = ("The '%(key)s' option (with value '%(value)s') is"
+                   " deprecated, please replace it with '%(replacement_key)s'."
+                   " This option will become invalid in version"
+                   " %(invalidation_version)s.")
```

Comment:
We don't end logging messages with period.

Reply:
This warning is also used as the exception message and the previous implementation had a period as well so I was trying to be consistent. I can remove it.